recipes_source/distributed_device_mesh.rst

```diff
@@ -0,0 +1,135 @@
+.. create inter-node and intra-node process groups without worrying about how to set up ranks correctly for different sub process groups.
+.. Users can also easily manage the underlying process_groups/devices for multi-dimensional parallelism via :class:`DeviceMesh`.
+
+.. figure:: /_static/img/distributed/device_mesh.png
+   :width: 100%
+   :align: center
+   :alt: PyTorch DeviceMesh
+
+Why DeviceMesh is Useful
+------------------------
+
+.. Below is the code snippet for a 2D setup without :class:`DeviceMesh`. First, we need to manually calculate shard group and replicate group. Then, we need to assign the correct shard and
+.. replicate group to each rank.
+
+.. code-block:: python
+import os
+
+import torch
+import torch.distributed as dist
+
+# Understand world topology
+rank = int(os.environ["RANK"])
+world_size = int(os.environ["WORLD_SIZE"])
+print(f"Running example on {rank=} in a world with {world_size=}")
+
+# Create process groups to manage 2-D like parallel pattern
+dist.init_process_group("nccl")
+
+# Create shard groups (e.g. (0, 1, 2, 3), (4, 5, 6, 7))
+# and assign the correct shard group to each rank
+num_node_devices = torch.cuda.device_count()
+shard_rank_lists = list(range(0, num_node_devices // 2)), list(range(num_node_devices // 2, num_node_devices))
+shard_groups = (
+    dist.new_group(shard_rank_lists[0]),
+    dist.new_group(shard_rank_lists[1]),
+)
+current_shard_group = (
+    shard_groups[0] if rank in shard_rank_lists[0] else shard_groups[1]
+)
+
+# Create replicate groups (e.g. (0, 4), (1, 5), (2, 6), (3, 7))
+# and assign the correct replicate group to each rank
+current_replicate_group = None
+shard_factor = len(shard_rank_lists[0])
+for i in range(num_node_devices // 2):
+    replicate_group_ranks = list(range(i, num_node_devices, shard_factor))
+    replicate_group = dist.new_group(replicate_group_ranks)
+    if rank in replicate_group_ranks:
+        current_replicate_group = replicate_group
+
+.. To run the above code snippet, we can leverage PyTorch Elastic. Let's create a file named ``2d_setup.py``.
+.. Then, run the following `torch elastic/torchrun <https://pytorch.org/docs/stable/elastic/quickstart.html>`__ command.
+
+.. code-block:: python
+torchrun --nnodes=1 --nproc_per_node=8 --rdzv_id=100 --rdzv_endpoint=localhost:29400 2d_setup.py
+
+
+.. With the help of :func:`init_device_mesh`, we can accomplish the above 2D setup in just 2 lines.
+
+
+.. code-block:: python
+from torch.distributed.device_mesh import init_device_mesh
+device_mesh = init_device_mesh("cuda", (2, 4))
+
+.. Let's create a file named ``2d_setup_with_device_mesh.py``.
+.. Then, run the following `torch elastic/torchrun <https://pytorch.org/docs/stable/elastic/quickstart.html>`__ command.
+
+.. code-block:: python
+torchrun --nnodes=1 --nproc_per_node=8 --rdzv_id=100 --rdzv_endpoint=localhost:29400 2d_setup_with_device_mesh.py
+
+
+How to use DeviceMesh with HSDP
+-------------------------------
+
+Hybrid Sharding(HSDP)
```

Comment:
Would like more info about HSDP. What it is what's the benefits/use cases of using with DeviceMesh, etc.